src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,221 @@
+                     Settings options,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Error,
+                              Var<Entry>,
+                              std::map<std::string, std::vector<std::string>>,
+                              Settings,
+                              Var<Reactor>,
+                              Var<Logger>> cb
+                    ) {
+    Var<std::map<std::string, std::vector<std::string>>>
+        fb_service_ips(new std::map<std::string, std::vector<std::string>>(
+            {{"stun", {}},
+             {"b_api", {}},
+             {"b_graph", {}},
+             {"edge", {}},
+             {"external_cdn", {}},
+             {"scontent_cdn", {}},
+             {"star", {}}}
+        ));
+
+    if (error) {
+        cb(error, entry, *fb_service_ips, options, reactor, logger);
+        return;
+    }
+
+    // if we find any inconsistent DNS later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int names_count = fb_service_hostnames.size();
+    Var<int> names_tested(new int(0));
+
+    if (names_count == *names_tested) {
+        cb(NoError(), entry, *fb_service_ips, options, reactor, logger);
+        return;
+    }
+
+    auto dns_cb = [=](std::string service, std::string hostname) {
+        return [=](Error err, Var<dns::Message> message) {
+            if (!!err) {
+                logger->info("fb_messenger: dns error for %s, %s",
+                    service.c_str(), hostname.c_str());
+            } else {
+                for (auto answer : message->answers) {
+                    if ((answer.ipv4 != "") || (answer.hostname != "")) {
+                        std::string asn_p = options.get("geoip_asn_path", std::string{});
+                        auto geoip = GeoipCache::thread_local_instance()->get(asn_p);
+                        ErrorOr<std::string> asn = geoip->resolve_asn(answer.ipv4);
+                        if (asn && asn.as_value() != "AS0") {
+                            logger->info("%s ipv4: %s, %s",
+                                hostname.c_str(), answer.ipv4.c_str(),
+                                asn.as_value().c_str());
+                            // if consistent, add to list for tcp connect later
+                            if (asn.as_value() == "AS32934") {
+                                (*fb_service_ips)[service].push_back(answer.ipv4);
+                            } else {
+                                (*entry)["facebook_dns_blocking"] = true;
+                            }
+                        }
+                    }
+                }
+                // if any consistent IPs, consistent = true
+                (*entry)["facebook_" + service + "_dns_consistent"] =
+                    !(*fb_service_ips)[service].empty();
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
```

Comment:
Here, can we `assert` that `names_count` is `<= *names_tested`. Also, can you declare `names_count` as `size_t` such that it cannot be negative? (Since you assign it using the `.size()` of a collection, `size_t` is the proper type anyway)